test/extensions/filters/http/lua/lua_filter_test.cc

```diff
@@ -1800,6 +1800,43 @@ TEST_F(LuaHttpFilterTest, GetRequestedServerName) {
   EXPECT_EQ(Http::FilterHeadersStatus::Continue, filter_->decodeHeaders(request_headers, true));
 }
 
+// Verify that extracting dynamicMetadata() using LUA also works for binary values.
+TEST_F(LuaHttpFilterTest, DynamicMetadataBinaryData) {
+  const std::string SCRIPT{R"EOF(
+    function envoy_on_request(request_handle)
+      local bin_data = request_handle:streamInfo():dynamicMetadata():get("envoy.pp")["bin_data"]
+      local hex_table = { }
+      local data_length = 8
+      for idx = 1, data_length do
+        local hex_value = string.format('%x', string.byte(bin_data, idx))
+        if string.len(string.format('%x', string.byte(bin_data, idx))) < 2 then
+          hex_table[data_length - idx + 1] = '0'..hex_value
+        else
+          hex_table[data_length - idx + 1] = hex_value
+        end
+      end
+      local hex_data = table.concat(hex_table, '')
+      local int64_data = string.format('%d', tonumber(hex_data, 16))
+      request_handle:logTrace('Data: ' .. int64_data)
+    end
+  )EOF"};
+
+  ProtobufWkt::Value metadata_value;
+  const uint64_t e0 = 9688686178336770;
+  metadata_value.set_string_value(reinterpret_cast<char const*>(&e0), 8);
```

Comment:
This is a bit too opaque about what's going on. The thing we really care about is that c_str() has embedded nulls, so let's write some of those. e.g.
constexpr uint8_t buffer[] = {'h', 'e', 0x00, 'l', 'l', 'o'};
metadata_value.set_string_value(buffer, sizeof(buffer));

maybe you need to throw in a reinterpret_cast in there, I didn't test this. And throw in a comment in the test that the prior version would result in only "he" being stored.